keanu-python/keanu/algorithm/sampling.py

```diff
@@ -17,31 +16,90 @@
+    def get_sampler(self) -> JavaObject:
+        return self._sampler
+
+
+class MetropolisHastingsSampler(PosteriorSamplingAlgorithm):
+
+    def __init__(self,
+                 proposal_distribution: str = None,
+                 proposal_distribution_sigma: numpy_types = None,
+                 proposal_listeners=[],
+                 use_cache_on_rejection: bool = None):
+
+        if (proposal_distribution is None and len(proposal_listeners) > 0):
+            raise TypeError("If you pass in proposal_listeners you must also specify proposal_distribution")
+
+        builder: JavaObject = k.jvm_view().MetropolisHastings.builder()
+
+        if proposal_distribution is not None:
+            proposal_distribution_object = ProposalDistribution(
+                type_=proposal_distribution, sigma=proposal_distribution_sigma, listeners=proposal_listeners)
+            builder = builder.proposalDistribution(proposal_distribution_object.unwrap())
+
+        if use_cache_on_rejection is not None:
+            builder.useCacheOnRejection(use_cache_on_rejection)
+
+        super().__init__(builder.build())
+
+
+class HamiltonianSampler(PosteriorSamplingAlgorithm):
+
+    def __init__(self) -> None:
+        super().__init__(k.jvm_view().Hamiltonian.withDefaultConfig())
+
+
+class NUTSSampler(PosteriorSamplingAlgorithm):
+
+    def __init__(self,
+                 adapt_count: int = None,
+                 target_acceptance_prob: float = None,
+                 adapt_enabled: bool = None,
+                 initial_step_size: float = None,
+                 max_tree_height: int = None):
+
+        builder: JavaObject = k.jvm_view().NUTS.builder()
+
+        if adapt_count is not None:
+            builder.adaptCount(adapt_count)
+
+        if target_acceptance_prob is not None:
+            builder.targetAcceptanceProb(target_acceptance_prob)
+
+        if adapt_enabled is not None:
+            builder.adaptEnabled(adapt_enabled)
+
+        if initial_step_size is not None:
+            builder.initialStepSize(initial_step_size)
+
+        if max_tree_height is not None:
+            builder.maxTreeHeight(max_tree_height)
+
+        super().__init__(builder.build())
 
 
 def sample(net: BayesNet,
            sample_from: Iterable[Vertex],
-           algo: str = 'metropolis',
-           proposal_distribution: str = None,
-           proposal_distribution_sigma: numpy_types = None,
-           proposal_listeners=[],
+           sampling_algorithm: PosteriorSamplingAlgorithm = None,
            draws: int = 500,
            drop: int = 0,
            down_sample_interval: int = 1,
            plot: bool = False,
            ax: Any = None) -> sample_types:
 
-    sampling_algorithm: JavaObject = build_sampling_algorithm(algo, proposal_distribution, proposal_distribution_sigma,
-                                                              proposal_listeners)
+    if sampling_algorithm is None:
+        sampling_algorithm = MetropolisHastingsSampler()
```

Comment:
add this to the `sample` method?

Reply:
Only if we want to change the semantics and share the "default" MetropolisHastingsSampler across all calls (which confuses some of the tests).  Also, it's generally considered bad form to create "mutable" objects as part of a default parameter in Python so I'd prefer to keep as is.